pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,230 @@
+package clusterprovisioner
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/helm/helm-mapkubeapis/pkg/mapping"
+	"github.com/pkg/errors"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/mod/semver"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/release"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	apiMappings = &mapping.Metadata{
+		Mappings: []*mapping.Mapping{
+			{
+				DeprecatedAPI:    "apiVersion: policy/v1beta1\nkind: PodSecurityPolicy\n",
+				RemovedInVersion: "v1.25",
+				// TODO add all APIs we want to cover here. Perhaps store it in a ConfigMap so we don't need to change
+				// code for updates.
+			},
+		},
+	}
+)
+
+type ClientGetter struct {
+	k8sClient  kubernetes.Clientset
+	restConfig rest.Config
+}
+
+func (c ClientGetter) ToRESTConfig() (*rest.Config, error) {
+	return &c.restConfig, nil
+}
+
+func (c ClientGetter) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
+	return memory.NewMemCacheClient(c.k8sClient.Discovery()), nil
+}
+
+func (c ClientGetter) ToRESTMapper() (meta.RESTMapper, error) {
+	return meta.NewDefaultRESTMapper(nil), nil
+}
+
+func (c ClientGetter) ToRawKubeConfigLoader() clientcmd.ClientConfig {
+	return &clientcmd.DefaultClientConfig
+}
+
+func NewClientGetter(k8sClient kubernetes.Interface, restConfig rest.Config) ClientGetter {
+	return ClientGetter{
+		k8sClient:  *k8sClient.(*kubernetes.Clientset),
+		restConfig: restConfig,
+	}
+}
+
+func (p *Provisioner) cleanupHelmReleases(cluster *v3.Cluster) error {
+	clusterManager := p.clusterManager
+	client, err := clusterManager.K8sClient(cluster.Name)
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to obtain the Kubernetes client instance")
+	}
+
+	restConfig, err := clusterManager.RESTConfig(cluster.Name)
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to obtain the Kubernetes REST config instance")
+	}
+
+	listNamespaces, err := client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
```

Comment:
Fair enough.

Reply:
Side question: I see you raised a concern about this potentially breaking charts we don't control. How do you feel about changing this to target solely Rancher-controlled namespaces?

Another idea would be feature gating it to target all namespaces if and only if a cluster admin chooses to do so.

Reply:
Big fan of only doing this for rancher controlled namespaces - I think that's a good idea.

I'm a bit shaky on allowing this to be configurable for other namespaces, partially because we need a way to expose the options to our users. If we used a configmap users would have to know it's there so that's substantial extra docs. At that point, users also probably should just upgrade their apps properly or run the command themselves. In short, given the extra work and the fact that users should probably be handling their own namespaces by themselves, I'd be more in favor of not allowing this to run on non-rancher namespaces.

Reply:
This change I've made actually very much breaks due to an import cycle as well.

@jiaqiluo if I intend to consume the `FeatureAppsNS` slice here, is there a different place where it could be stored that does not generate an import cycle and does not break your logic either?

Reply:
Hmm, I do not know a good place for it, but you are welcome to move and even rename the variable to fit your needs.

Reply:
@jiaqiluo I created another package and moved it there, could you please take a look and see if you agree? Thanks!

Reply:
The new package LGMT.  Good job!